File: sketch/src/main/java/me/xiaopan/sketch/decode/NormalDecodeHelper.java
/*
 * Copyright (C) 2017 Peng fei Pan <sky@xiaopan.me>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package me.xiaopan.sketch.decode;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;

import me.xiaopan.sketch.ErrorTracker;
import me.xiaopan.sketch.SLog;
import me.xiaopan.sketch.SLogType;
import me.xiaopan.sketch.cache.BitmapPool;
import me.xiaopan.sketch.cache.BitmapPoolUtils;
import me.xiaopan.sketch.drawable.ImageAttrs;
import me.xiaopan.sketch.request.LoadRequest;
import me.xiaopan.sketch.request.MaxSize;
import me.xiaopan.sketch.util.SketchUtils;

public class NormalDecodeHelper extends DecodeHelper {
    private static final String LOG_NAME = "NormalDecodeHelper";

    @Override
    public boolean match(LoadRequest request, DataSource dataSource, ImageType imageType,
                         BitmapFactory.Options boundOptions) {
        return true;
    }

    @Override
    public DecodeResult decode(LoadRequest request, DataSource dataSource, ImageType imageType,
                               BitmapFactory.Options boundOptions, BitmapFactory.Options decodeOptions, int exifOrientation) throws DecodeException {

        ImageOrientationCorrector orientationCorrector = request.getConfiguration().getImageOrientationCorrector();
        orientationCorrector.rotateSize(boundOptions, exifOrientation);

        // Calculate inSampleSize according to max size
        MaxSize maxSize = request.getOptions().getMaxSize();
        if (maxSize != null) {
            boolean supportLargeImage = SketchUtils.supportLargeImage(request, imageType);
            ImageSizeCalculator imageSizeCalculator = request.getConfiguration().getImageSizeCalculator();
            decodeOptions.inSampleSize = imageSizeCalculator.calculateInSampleSize(boundOptions.outWidth, boundOptions.outHeight,
                    maxSize.getWidth(), maxSize.getHeight(), supportLargeImage);
        }

        // Set inBitmap from bitmap pool
        if (BitmapPoolUtils.sdkSupportInBitmap() && !request.getOptions().isBitmapPoolDisabled()) {
            BitmapPool bitmapPool = request.getConfiguration().getBitmapPool();
            BitmapPoolUtils.setInBitmapFromPool(decodeOptions,
                    boundOptions.outWidth, boundOptions.outHeight, boundOptions.outMimeType, bitmapPool);
        }

        Bitmap bitmap = null;
        try {
            bitmap = ImageDecodeUtils.decodeBitmap(dataSource, decodeOptions);
        } catch (Throwable throwable) {
            throwable.printStackTrace();

            ErrorTracker errorTracker = request.getConfiguration().getErrorTracker();
            BitmapPool bitmapPool = request.getConfiguration().getBitmapPool();
            if (ImageDecodeUtils.isInBitmapDecodeError(throwable, decodeOptions, false)) {
                ImageDecodeUtils.recycleInBitmapOnDecodeError(errorTracker, bitmapPool, request.getUri(),
                        boundOptions.outWidth, boundOptions.outHeight, boundOptions.outMimeType, throwable, decodeOptions, false);

                try {
                    bitmap = ImageDecodeUtils.decodeBitmap(dataSource, decodeOptions);
                } catch (Throwable throwable1) {
                    throwable1.printStackTrace();

                    errorTracker.onDecodeNormalImageError(throwable1, request,
                            boundOptions.outWidth, boundOptions.outHeight, boundOptions.outMimeType);
                }
            } else {
                errorTracker.onDecodeNormalImageError(throwable, request,
                        boundOptions.outWidth, boundOptions.outHeight, boundOptions.outMimeType);
            }
        }

        // 过滤掉无效的图片
        if (bitmap == null || bitmap.isRecycled()) {
            ImageDecodeUtils.decodeError(request, dataSource, LOG_NAME);
            return null;
        }

        // 过滤宽高小于等于1的图片
        if (bitmap.getWidth() <= 1 || bitmap.getHeight() <= 1) {
            SLog.fw(SLogType.REQUEST, LOG_NAME,
                    "image width or height less than or equal to 1px. imageSize: %dx%d. bitmapSize: %dx%d. %s",
                    boundOptions.outWidth, boundOptions.outHeight, bitmap.getWidth(), bitmap.getHeight(), request.getKey());
            bitmap.recycle();
            ImageDecodeUtils.decodeError(request, dataSource, LOG_NAME);
            return null;
        }

        ProcessedImageCache processedImageCache = request.getConfiguration().getProcessedImageCache();
        boolean processed = processedImageCache.canUseCacheProcessedImageInDisk(decodeOptions.inSampleSize);

        ImageAttrs imageAttrs = new ImageAttrs(boundOptions.outMimeType, boundOptions.outWidth, boundOptions.outHeight, exifOrientation);
        BitmapDecodeResult result = new BitmapDecodeResult(imageAttrs, bitmap).setProcessed(processed);

        correctOrientation(orientationCorrector, result, exifOrientation, request);

        ImageDecodeUtils.decodeSuccess(bitmap, boundOptions.outWidth, boundOptions.outHeight, decodeOptions.inSampleSize, request, LOG_NAME);
        return result;
    }
}

File: sketch3/src/main/java/com/github/panpf/sketch/common/decode/internal/NormalDecodeHelper.kt
/*
 * Copyright (C) 2019 panpf <panpfpanpf@outlook.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.github.panpf.sketch.common.decode.internal

import android.graphics.Bitmap
import android.graphics.BitmapFactory.Options
import android.graphics.Point
import com.github.panpf.sketch.SLog
import com.github.panpf.sketch.Sketch
import com.github.panpf.sketch.common.DecodeException
import com.github.panpf.sketch.common.LoadableRequest
import com.github.panpf.sketch.common.cache.BitmapPoolHelper
import com.github.panpf.sketch.common.datasource.DataSource
import com.github.panpf.sketch.common.decode.DecodeResult
import com.github.panpf.sketch.load.ImageInfo

class NormalDecodeHelper {

    companion object {
        private const val MODULE = "NormalDecodeHelper"
    }

    private val sizeCalculator = ImageSizeCalculator()

    fun decode(
        sketch: Sketch,
        request: LoadableRequest,
        dataSource: DataSource,
        imageInfo: ImageInfo,
        decodeOptions: Options,
        imageOrientationCorrector: ImageOrientationCorrector?
    ): Bitmap {
        val resize = request.resize
        val maxSize = request.maxSize
        val bitmapPoolHelper = sketch.bitmapPoolHelper
        val imageSize = Point(imageInfo.width, imageInfo.height)
        imageOrientationCorrector?.rotateSize(imageSize)

        val maxSizeInSampleSize = maxSize?.let {
            sizeCalculator.calculateInSampleSize(
                imageSize.x, imageSize.y, it.width, it.height
            )
        } ?: 1
        val resizeInSampleSize = resize?.let {
            sizeCalculator.calculateInSampleSize(
                imageSize.x, imageSize.y, it.width, it.height
            )
        } ?: 1
        decodeOptions.inSampleSize = maxSizeInSampleSize.coerceAtLeast(resizeInSampleSize)

        // Set inBitmap from bitmap pool
        if (request.disabledBitmapPool != true) {
            bitmapPoolHelper
                .setInBitmap(decodeOptions, imageSize.x, imageSize.y, imageInfo.mimeType)
        }

        return decodeBitmap(request, dataSource, decodeOptions, bitmapPoolHelper)
    }

    private fun decodeBitmap(
        request: LoadableRequest,
        dataSource: DataSource,
        decodeOptions: Options,
        bitmapPoolHelper: BitmapPoolHelper,
    ): Bitmap {
        val bitmap: Bitmap? = try {
            dataSource.decodeBitmap(decodeOptions)
        } catch (throwable: Throwable) {
            val inBitmap = decodeOptions.inBitmap
            if (inBitmap != null && isInBitmapError(throwable, false)) {
                val message = "Bitmap decode error. Because inBitmap. uri=%s"
                    .format(request.uri)
                SLog.emt(MODULE, throwable, message)

                decodeOptions.inBitmap = null
                bitmapPoolHelper.freeBitmapToPool(inBitmap)
                try {
                    dataSource.decodeBitmap(decodeOptions)
                } catch (throwable2: Throwable) {
                    val message2 = "Bitmap decode error. uri=%s".format(request.uri)
                    SLog.emt(MODULE, throwable2, message2)
                    throw DecodeException(message2, throwable2)
                }
            } else {
                val message = "Bitmap decode error. uri=%s".format(request.uri)
                SLog.emt(MODULE, throwable, message)
                throw DecodeException(message, throwable)
            }
        }
        if (bitmap == null) {
            val message = "Bitmap decode return null. uri=%s".format(request.uri)
            SLog.em(MODULE, message)
            throw DecodeException(message)
        }

        if (bitmap.width <= 1 || bitmap.height <= 1) {
            bitmap.recycle()
            val message = "Invalid image size. size=%dx%d, uri=%s"
                .format(bitmap.width, bitmap.height, request.uri)
            SLog.em(MODULE, message)
            throw DecodeException(message)
        }

        return bitmap
    }
}